src/dune_rules/warning.ml

```diff
@@ -0,0 +1,78 @@
+open Import
+
+module Name = struct
+  include String
+
+  include Stringlike.Make (struct
+      type t = string
+
+      let module_ = "Warning"
+      let description = "Warning name"
+      let description_of_valid_string = None
+      let hint_valid = None
+      let of_string_opt x = if Dune_sexp.Atom.is_valid x then Some x else None
+      let to_string x = x
+    end)
+end
+
+type t =
+  { name : Name.t
+  ; since : Syntax.Version.t
+  ; default : Syntax.Version.t -> Config.Toggle.t
+  }
+
+let frozen = ref false
+let all : (Name.t, t) Table.t = Table.create (module Name) 12
+
+let create ~default ~name ~since =
+  if !frozen then Code_error.raise "Warnings may not be created after we've read them" [];
```

Comment:
Can you add the warning name to the list passed to `Code_error.raise`. Also if I ran into this case as a developer I'd find it more helpful if the message said something like: `"Warning.create may not be called after warning settings have been parsed. Warnings should be created at the top level (not inside function definitions) in the file src/dune_rules/warnings.ml"`.